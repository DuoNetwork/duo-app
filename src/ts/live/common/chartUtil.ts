import moment from 'moment';
import * as CST from '../common/constants';
import { IAcceptedPrice, ICustodianPrice, ICustodianStates, IPriceBar, ISourceData } from './types';

class ChartUtil {
	public interpolate(sourceData: IPriceBar[], isHourly: boolean): IPriceBar[] {
		if (!sourceData.length) return [];
		sourceData.sort((a, b) => a.timestamp - b.timestamp);
		const newSourceData: IPriceBar[] = [];
		const timeStep = isHourly ? 3.6e6 : 60000;
		const source = sourceData[0].source;
		for (let i = 0; i < sourceData.length - 1; i++) {
			newSourceData.push(sourceData[i]);
			const timeIndex = sourceData[i].timestamp;
			const nextTimeIndex = sourceData[i + 1].timestamp;
			const close = sourceData[i].close;
			let newTimeStamp = timeIndex + timeStep;
			while (newTimeStamp < nextTimeIndex) {
				newSourceData.push({
					source: source,
					date: moment(newTimeStamp).format('YYYY-MM-DD'),
					hour: moment(newTimeStamp).format('HH'),
					minute: isHourly ? 0 : Number(moment(newTimeStamp).format('mm')),
					open: close,
					high: close,
					low: close,
					close: close,
					volume: 0,
					timestamp: newTimeStamp
				});
				newTimeStamp += timeStep;
			}
		}
		newSourceData.push(sourceData[sourceData.length - 1]);

		return newSourceData;
	}

	public reset(
		price: IAcceptedPrice[],
		limitUp: number,
		limitDown: number,
		limitPeriod: number
	): IAcceptedPrice[] {
		return price
			.filter(p => p.navB >= limitUp || p.navB <= limitDown || p.navA >= limitPeriod)
			.map(p => ({
				price: p.price,
				navA: 1,
				navB: 1,
				timestamp: p.timestamp + 717
			}));
	}

	public mergeReset(prices: IAcceptedPrice[], resets: IAcceptedPrice[]): IAcceptedPrice[] {
		const all = [...prices, ...resets];
		all.sort((a, b) => a.timestamp - b.timestamp);
		return all;
	}

	public mergeLastToPriceBar(
		pricebars: ISourceData<IPriceBar[]>,
		last: ISourceData<ICustodianPrice>,
		isHourly: boolean
	) {
		CST.EXCHANGES.forEach(src => {
			const pricebar: IPriceBar[] = pricebars[src.toLowerCase()];
			if (!pricebar || !pricebar.length) return;
			const sourceLast: ICustodianPrice = last[src.toLowerCase()];
			const lastBar = pricebar[pricebar.length - 1];
			const dateObj = moment.utc(sourceLast.timestamp);
			if (isHourly && lastBar.date + ' ' + lastBar.hour !== dateObj.format('YYYY-MM-DD HH'))
				pricebar.push({
					source: lastBar.source,
					date: dateObj.format('YYYY-MM-DD'),
					hour: dateObj.format('HH'),
					minute: 0,
					open: lastBar.close,
					high: Math.max(lastBar.close, sourceLast.price),
					low: Math.min(lastBar.close, sourceLast.price),
					close: sourceLast.price,
					volume: 0,
					timestamp: sourceLast.timestamp
				});
			else if (
				!isHourly &&
				lastBar.date + ' ' + lastBar.hour + ' ' + lastBar.minute !==
					dateObj.format('YYYY-MM-DD HH m')
			)
				pricebar.push({
					source: lastBar.source,
					date: dateObj.format('YYYY-MM-DD'),
					hour: dateObj.format('HH'),
					minute: Number(dateObj.format('mm')),
					open: lastBar.close,
					high: Math.max(lastBar.close, sourceLast.price),
					low: Math.min(lastBar.close, sourceLast.price),
					close: sourceLast.price,
					volume: 0,
					timestamp: sourceLast.timestamp
				});
			else {
				lastBar.high = Math.max(lastBar.high, sourceLast.price);
				lastBar.low = Math.min(lastBar.low, sourceLast.price);
				lastBar.close = sourceLast.price;
				lastBar.timestamp = sourceLast.timestamp;
			}
		});
	}

	public mergeLastToPrice(
		prices: IAcceptedPrice[],
		states: ICustodianStates,
		last: ICustodianPrice
	) {
		if (!prices.length) return;
		const lastPrice = prices[prices.length - 1];
		if (lastPrice.timestamp < last.timestamp)
			prices.push({
				price: last.price,
				navA: states.navA,
				navB: states.navB,
				timestamp: last.timestamp
			});
	}
}

const chartUtil = new ChartUtil();
export default chartUtil;
