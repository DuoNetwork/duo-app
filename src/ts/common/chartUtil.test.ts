import moment from 'moment';
// import hourly from '../samples/hourly.json';
// import minutely from '../samples/minutely.json';
// import prices from '../samples/prices.json';
import chartUtil from './chartUtil';
// import util from './util';

const hourlyBar1 = {
	source: 'source',
	base: 'base',
	quote: 'quote',
	period: 60,
	open: 1,
	high: 3,
	low: 0,
	close: 2,
	volume: 123,
	timestamp: moment.utc('2018-06-06 00:0', 'YYYY-MM-DD HH:m').valueOf()
};

const hourlyBar2 = {
	source: 'source',
	base: 'base',
	quote: 'quote',
	period: 60,
	open: 1,
	high: 3,
	low: 0,
	close: 2,
	volume: 123,
	timestamp: moment.utc('2018-06-06 01:0', 'YYYY-MM-DD HH:m').valueOf()
};

const hourlyBar3 = {
	source: 'source',
	base: 'base',
	quote: 'quote',
	period: 60,
	open: 1,
	high: 3,
	low: 0,
	close: 2,
	volume: 123,
	timestamp: moment.utc('2018-06-06 03:0', 'YYYY-MM-DD HH:m').valueOf()
};

test('mergePrices', () => {
	expect(chartUtil.mergePrices([hourlyBar1, hourlyBar2, hourlyBar3], 180)).toMatchSnapshot();
});

const minutelyBar1 = {
	source: 'source',
	base: 'base',
	quote: 'quote',
	period: 1,
	open: 1,
	high: 3,
	low: 0,
	close: 2,
	volume: 123,
	timestamp: moment.utc('2018-06-06 00:0', 'YYYY-MM-DD HH:m').valueOf()
};

const minutelyBar2 = {
	source: 'source',
	base: 'base',
	quote: 'quote',
	period: 1,
	open: 1,
	high: 3,
	low: 0,
	close: 2,
	volume: 123,
	timestamp: moment.utc('2018-06-06 00:1', 'YYYY-MM-DD HH:m').valueOf()
};

const minutelyBar3 = {
	source: 'source',
	base: 'base',
	quote: 'quote',
	period: 1,
	open: 1,
	high: 3,
	low: 0,
	close: 2,
	volume: 123,
	timestamp: moment.utc('2018-06-06 00:3', 'YYYY-MM-DD HH:m').valueOf()
};

test('minutely merged correctly to 5 minute bar', () => {
	expect(chartUtil.mergePrices([minutelyBar1, minutelyBar2, minutelyBar3], 5)).toMatchSnapshot();
});

const acceptedPrice1 = {
	contractAddress: '0x0',
	price: 123,
	navA: 1,
	navB: 1,
	timestamp: 1000,
	blockNumber: 111,
	transactionHash: 'aaa'
};

const acceptedPrice2 = {
	contractAddress: '0x0',
	price: 123,
	navA: 1,
	navB: 2,
	timestamp: 2000,
	blockNumber: 222,
	transactionHash: 'bbb'
};

const acceptedPrice3 = {
	contractAddress: '0x0',
	price: 123,
	navA: 1,
	navB: 0.25,
	timestamp: 3000,
	blockNumber: 333,
	transactionHash: 'ccc'
};

const acceptedPrice4 = {
	contractAddress: '0x0',
	price: 123,
	navA: 1.1,
	navB: 1,
	timestamp: 4000,
	blockNumber: 444,
	transactionHash: 'ddd'
};

const acceptedPriceNow = {
	contractAddress: '0x0',
	price: 123,
	navA: 1.1,
	navB: 1,
	timestamp: (Number(moment.now.valueOf()) / 1000),
	blockNumber: 444,
	transactionHash: 'ddd'
};

const reset = chartUtil.reset(
	[acceptedPrice1, acceptedPrice2, acceptedPrice3, acceptedPrice4],
	2,
	0.25,
	1.05
);

const beethovenStates = {
	lastOperationTime: 12,
	operationCoolDown: 23,
	alpha: 12,
	beta: 23,
	createCommRate: 0.01,
	duoBalance: 2482128.9762267205,
	ethBalance: 742.408092589115,
	ethDuoFeeRatio: 800,
	feeAccumulated: 87.06788278541666,
	iterationGasThreshold: 65000,
	lastAdminTime: 0,
	limitLower: 0.25,
	limitPeriodic: 1.035,
	limitUpper: 2,
	navA: 1.025942,
	navB: 0.595738272490127,
	nextResetAddrIndex: 0,
	numOfPrices: 0,
	period: 3600,
	periodCoupon: 0.000017,
	preResetWaitingBlocks: 10,
	priceFeedTimeTol: 60,
	priceFeedTol: 0.01,
	priceTol: 0.05,
	priceUpdateCoolDown: 3000,
	redeemCommRate: 0.01,
	state: 'Trading',
	totalSupplyA: 82880.4753362674,
	totalSupplyB: 82880.4753362674,
	usersLength: 295,
	minBalance: 6,
	ethCollateral: 6,
	lastPrice: 6,
	lastPriceTime: 6,
	resetPrice: 6,
	resetPriceTime: 6,
	priceFetchCoolDown: 6,
	totalUsers: 6,
	feeBalance: 6,
	resetState: 'true',
	maturity: 12,
};

const contractPrice = {
	price: 12,
	timestamp: moment.utc('2018-06-06 00:0', 'YYYY-MM-DD HH:m').valueOf()
};

test('find reset points correctly', () => expect(reset).toMatchSnapshot());

test('merge reset correctly', () =>
	expect(
		chartUtil.mergeReset(
			[acceptedPrice1, acceptedPrice2, acceptedPrice3, acceptedPrice4],
			reset
		)
	).toMatchSnapshot());

test('merge Prices correctly', () => {
	expect(chartUtil.mergePricesToPrice([hourlyBar1, hourlyBar2, hourlyBar3], 5)).toMatchSnapshot();
	expect(chartUtil.mergePricesToPrice([], 5)).toMatchSnapshot();
});

test('merge last price correctly', () => {
	const lastDataBefore = { address: '0x0', price: 205.01, timestamp: 1528200000 };
	const lastData = { address: '0x0', price: 205.01, timestamp: 1539755124695 };
	const lastDataNew = { address: '0x0', price: 205.01, timestamp: 1528243212000 };
	const lastDataSameTime = {
		address: '0x0',
		price: 205.1,
		timestamp: moment.utc('2018-06-06 00:0', 'YYYY-MM-DD HH:m').valueOf()
	};
	expect(chartUtil.mergeLastestToPrices([], lastData)).toMatchSnapshot();
	expect(chartUtil.mergeLastestToPrices([], lastDataBefore)).toMatchSnapshot();
	expect(
		chartUtil.mergeLastestToPrices([hourlyBar1, hourlyBar2, hourlyBar3], lastData)
	).toMatchSnapshot();
	expect(
		chartUtil.mergeLastestToPrices([hourlyBar1, hourlyBar2, hourlyBar3], lastDataNew)
	).toMatchSnapshot();
	expect(
		chartUtil.mergeLastestToPrices([hourlyBar1, hourlyBar2, hourlyBar3], lastDataBefore)
	).toMatchSnapshot();
	expect(
		chartUtil.mergeLastestToPrices([hourlyBar1, hourlyBar2, hourlyBar3], lastDataSameTime)
	).toMatchSnapshot();
});

test('mergeLastestToAcceptedPrices', () => {
	expect(
		chartUtil.mergeLastestToAcceptedPrices(
			[acceptedPrice1, acceptedPrice2, acceptedPrice3, acceptedPrice4],
			beethovenStates,
			contractPrice
		)
	).toMatchSnapshot();
	expect(
		chartUtil.mergeLastestToAcceptedPrices([], beethovenStates, contractPrice)
	).toMatchSnapshot();
	expect(
		chartUtil.mergeLastestToAcceptedPrices(
			[acceptedPrice1, acceptedPrice2, acceptedPrice3, acceptedPrice4, acceptedPriceNow],
			beethovenStates,
			contractPrice
		)
	).toMatchSnapshot();
});

// test('merge last price to accepted price correctly', () => {
// 	const lastData = { address: '0x0', price: 205.01, timestamp: 1539755124695 };
// 	const lastDataSameTime = { address: '0x0', price: 250.01, timestamp: 4000 };
// 	const states = {
// 		addrPoolLength: 6,
// 		adminCoolDown: 3600,
// 		alpha: 1,
// 		beta: 1,
// 		createCommRate: 0.01,
// 		duoBalance: 2482128.9762267205,
// 		ethBalance: 742.408092589115,
// 		ethDuoFeeRatio: 800,
// 		feeAccumulated: 87.06788278541666,
// 		iterationGasThreshold: 65000,
// 		lastAdminTime: 0,
// 		limitLower: 0.25,
// 		limitPeriodic: 1.035,
// 		limitUpper: 2,
// 		navA: 1.025942,
// 		navB: 0.595738272490127,
// 		nextResetAddrIndex: 0,
// 		numOfPrices: 0,
// 		period: 3600,
// 		periodCoupon: 0.000017,
// 		preResetWaitingBlocks: 10,
// 		priceFeedTimeTol: 60,
// 		priceFeedTol: 0.01,
// 		priceTol: 0.05,
// 		priceUpdateCoolDown: 3000,
// 		redeemCommRate: 0.01,
// 		state: 'Trading',
// 		totalSupplyA: 82880.4753362674,
// 		totalSupplyB: 82880.4753362674,
// 		usersLength: 295
// 	};
// 	expect(
// 		chartUtil.mergeLastestToAcceptedPrices(
// 			[acceptedPrice1, acceptedPrice2, acceptedPrice3],
// 			states,
// 			lastData
// 		)
// 	).toMatchSnapshot();
// 	expect(chartUtil.mergeLastestToAcceptedPrices([], states, lastData)).toMatchSnapshot();
// 	expect(
// 		chartUtil.mergeLastestToAcceptedPrices(
// 			[acceptedPrice1, acceptedPrice2, acceptedPrice3],
// 			states,
// 			lastDataSameTime
// 		)
// 	).toMatchSnapshot();
// });
