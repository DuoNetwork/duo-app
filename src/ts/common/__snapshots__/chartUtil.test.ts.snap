// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`find reset points correctly 1`] = `
Array [
  Object {
    "blockNumber": 222,
    "contractAddress": "0x0",
    "navA": 1,
    "navB": 1,
    "price": 123,
    "timestamp": 2717,
    "transactionHash": "bbb",
  },
  Object {
    "blockNumber": 333,
    "contractAddress": "0x0",
    "navA": 1,
    "navB": 1,
    "price": 123,
    "timestamp": 3717,
    "transactionHash": "ccc",
  },
  Object {
    "blockNumber": 444,
    "contractAddress": "0x0",
    "navA": 1,
    "navB": 1,
    "price": 123,
    "timestamp": 4717,
    "transactionHash": "ddd",
  },
]
`;

exports[`merge Prices correctly 1`] = `
Object {
  "base": "base",
  "close": 2,
  "high": 3,
  "low": 0,
  "open": 1,
  "period": 5,
  "quote": "quote",
  "source": "source",
  "timestamp": 1528254000000,
  "volume": 369,
}
`;

exports[`merge Prices correctly 2`] = `
Object {
  "base": "",
  "close": 0,
  "high": 0,
  "low": 0,
  "open": 0,
  "period": 5,
  "quote": "",
  "source": "",
  "timestamp": 0,
  "volume": 0,
}
`;

exports[`merge last price correctly 1`] = `Array []`;

exports[`merge last price correctly 2`] = `Array []`;

exports[`merge last price correctly 3`] = `
Array [
  Object {
    "base": "base",
    "close": 205.01,
    "high": 205.01,
    "low": 2,
    "open": 2,
    "period": 60,
    "quote": "quote",
    "source": "source",
    "timestamp": 1539752400000,
    "volume": 0,
  },
  Object {
    "base": "base",
    "close": 2,
    "high": 3,
    "low": 0,
    "open": 1,
    "period": 60,
    "quote": "quote",
    "source": "source",
    "timestamp": 1528243200000,
    "volume": 123,
  },
  Object {
    "base": "base",
    "close": 2,
    "high": 3,
    "low": 0,
    "open": 1,
    "period": 60,
    "quote": "quote",
    "source": "source",
    "timestamp": 1528246800000,
    "volume": 123,
  },
  Object {
    "base": "base",
    "close": 2,
    "high": 3,
    "low": 0,
    "open": 1,
    "period": 60,
    "quote": "quote",
    "source": "source",
    "timestamp": 1528254000000,
    "volume": 123,
  },
]
`;

exports[`merge last price correctly 4`] = `
Array [
  Object {
    "base": "base",
    "close": 205.01,
    "high": 205.01,
    "low": 0,
    "open": 1,
    "period": 60,
    "quote": "quote",
    "source": "source",
    "timestamp": 1528243200000,
    "volume": 0,
  },
  Object {
    "base": "base",
    "close": 2,
    "high": 3,
    "low": 0,
    "open": 1,
    "period": 60,
    "quote": "quote",
    "source": "source",
    "timestamp": 1528246800000,
    "volume": 123,
  },
  Object {
    "base": "base",
    "close": 2,
    "high": 3,
    "low": 0,
    "open": 1,
    "period": 60,
    "quote": "quote",
    "source": "source",
    "timestamp": 1528254000000,
    "volume": 123,
  },
]
`;

exports[`merge last price correctly 5`] = `
Array [
  Object {
    "base": "base",
    "close": 2,
    "high": 3,
    "low": 0,
    "open": 1,
    "period": 60,
    "quote": "quote",
    "source": "source",
    "timestamp": 1528243200000,
    "volume": 123,
  },
  Object {
    "base": "base",
    "close": 2,
    "high": 3,
    "low": 0,
    "open": 1,
    "period": 60,
    "quote": "quote",
    "source": "source",
    "timestamp": 1528246800000,
    "volume": 123,
  },
  Object {
    "base": "base",
    "close": 2,
    "high": 3,
    "low": 0,
    "open": 1,
    "period": 60,
    "quote": "quote",
    "source": "source",
    "timestamp": 1528254000000,
    "volume": 123,
  },
]
`;

exports[`merge last price correctly 6`] = `
Array [
  Object {
    "base": "base",
    "close": 2,
    "high": 3,
    "low": 0,
    "open": 1,
    "period": 60,
    "quote": "quote",
    "source": "source",
    "timestamp": 1528243200000,
    "volume": 123,
  },
  Object {
    "base": "base",
    "close": 2,
    "high": 3,
    "low": 0,
    "open": 1,
    "period": 60,
    "quote": "quote",
    "source": "source",
    "timestamp": 1528246800000,
    "volume": 123,
  },
  Object {
    "base": "base",
    "close": 2,
    "high": 3,
    "low": 0,
    "open": 1,
    "period": 60,
    "quote": "quote",
    "source": "source",
    "timestamp": 1528254000000,
    "volume": 123,
  },
]
`;

exports[`merge reset correctly 1`] = `
Array [
  Object {
    "blockNumber": 111,
    "contractAddress": "0x0",
    "navA": 1,
    "navB": 1,
    "price": 123,
    "timestamp": 1000,
    "transactionHash": "aaa",
  },
  Object {
    "blockNumber": 222,
    "contractAddress": "0x0",
    "navA": 1,
    "navB": 2,
    "price": 123,
    "timestamp": 2000,
    "transactionHash": "bbb",
  },
  Object {
    "blockNumber": 222,
    "contractAddress": "0x0",
    "navA": 1,
    "navB": 1,
    "price": 123,
    "timestamp": 2717,
    "transactionHash": "bbb",
  },
  Object {
    "blockNumber": 333,
    "contractAddress": "0x0",
    "navA": 1,
    "navB": 0.25,
    "price": 123,
    "timestamp": 3000,
    "transactionHash": "ccc",
  },
  Object {
    "blockNumber": 333,
    "contractAddress": "0x0",
    "navA": 1,
    "navB": 1,
    "price": 123,
    "timestamp": 3717,
    "transactionHash": "ccc",
  },
  Object {
    "blockNumber": 444,
    "contractAddress": "0x0",
    "navA": 1.1,
    "navB": 1,
    "price": 123,
    "timestamp": 4000,
    "transactionHash": "ddd",
  },
  Object {
    "blockNumber": 444,
    "contractAddress": "0x0",
    "navA": 1,
    "navB": 1,
    "price": 123,
    "timestamp": 4717,
    "transactionHash": "ddd",
  },
]
`;

exports[`mergeLastestToAcceptedPrices 1`] = `
Array [
  Object {
    "blockNumber": 111,
    "contractAddress": "0x0",
    "navA": 1,
    "navB": 1,
    "price": 123,
    "timestamp": 1000,
    "transactionHash": "aaa",
  },
  Object {
    "blockNumber": 222,
    "contractAddress": "0x0",
    "navA": 1,
    "navB": 2,
    "price": 123,
    "timestamp": 2000,
    "transactionHash": "bbb",
  },
  Object {
    "blockNumber": 333,
    "contractAddress": "0x0",
    "navA": 1,
    "navB": 0.25,
    "price": 123,
    "timestamp": 3000,
    "transactionHash": "ccc",
  },
  Object {
    "blockNumber": 444,
    "contractAddress": "0x0",
    "navA": 1.1,
    "navB": 1,
    "price": 123,
    "timestamp": 4000,
    "transactionHash": "ddd",
  },
  Object {
    "blockNumber": 0,
    "contractAddress": "0x0",
    "navA": 1.025942,
    "navB": 0.595738272490127,
    "price": 12,
    "timestamp": 1528243200000,
    "transactionHash": "",
  },
]
`;

exports[`mergeLastestToAcceptedPrices 2`] = `Array []`;

exports[`mergeLastestToAcceptedPrices 3`] = `
Array [
  Object {
    "blockNumber": 111,
    "contractAddress": "0x0",
    "navA": 1,
    "navB": 1,
    "price": 123,
    "timestamp": 1000,
    "transactionHash": "aaa",
  },
  Object {
    "blockNumber": 222,
    "contractAddress": "0x0",
    "navA": 1,
    "navB": 2,
    "price": 123,
    "timestamp": 2000,
    "transactionHash": "bbb",
  },
  Object {
    "blockNumber": 333,
    "contractAddress": "0x0",
    "navA": 1,
    "navB": 0.25,
    "price": 123,
    "timestamp": 3000,
    "transactionHash": "ccc",
  },
  Object {
    "blockNumber": 444,
    "contractAddress": "0x0",
    "navA": 1.1,
    "navB": 1,
    "price": 123,
    "timestamp": 4000,
    "transactionHash": "ddd",
  },
  Object {
    "blockNumber": 444,
    "contractAddress": "0x0",
    "navA": 1.1,
    "navB": 1,
    "price": 123,
    "timestamp": NaN,
    "transactionHash": "ddd",
  },
]
`;

exports[`mergePrices 1`] = `
Array [
  Object {
    "base": "base",
    "close": 2,
    "high": 3,
    "low": 0,
    "open": 1,
    "period": 180,
    "quote": "quote",
    "source": "source",
    "timestamp": 1528243200000,
    "volume": 246,
  },
]
`;

exports[`minutely merged correctly to 5 minute bar 1`] = `
Array [
  Object {
    "base": "base",
    "close": 2,
    "high": 3,
    "low": 0,
    "open": 1,
    "period": 5,
    "quote": "quote",
    "source": "source",
    "timestamp": 1528243200000,
    "volume": 369,
  },
]
`;
